File: showcase/starters/langgraph-fastapi/src/agents/tools/sales_todos.py
"""Sales todos tool implementation."""

from __future__ import annotations

import uuid
from typing import Optional

from src.agents.types import SalesTodo

INITIAL_TODOS: list[SalesTodo] = [
    SalesTodo(
        id="st-001",
        title="Follow up with Acme Corp on enterprise proposal",
        stage="proposal",
        value=85000,
        dueDate="2026-04-15",
        assignee="Sarah Chen",
        completed=False,
    ),
    SalesTodo(
        id="st-002",
        title="Qualify lead from TechFlow demo request",
        stage="prospect",
        value=42000,
        dueDate="2026-04-18",
        assignee="Mike Johnson",
        completed=False,
    ),
    SalesTodo(
        id="st-003",
        title="Send contract to DataViz Inc for final review",
        stage="negotiation",
        value=120000,
        dueDate="2026-04-20",
        assignee="Sarah Chen",
        completed=False,
    ),
]

def manage_sales_todos_impl(todos: list[dict]) -> list[SalesTodo]:
    """Assign UUIDs to any todos missing an ID, then return the updated list."""
    result: list[SalesTodo] = []
    for todo in todos:
        result.append(SalesTodo(
            id=todo.get("id") or str(uuid.uuid4()),
            title=todo.get("title", ""),
            stage=todo.get("stage", "prospect"),
            value=todo.get("value", 0),
            dueDate=todo.get("dueDate", ""),
            assignee=todo.get("assignee", ""),
            completed=todo.get("completed", False),
        ))
    return result

def get_sales_todos_impl(current_todos: Optional[list[dict]] = None) -> list[SalesTodo]:
    """Return current todos or initial defaults if none provided."""
    if current_todos is not None:
        return manage_sales_todos_impl(current_todos)
    return list(INITIAL_TODOS)

File: showcase/integrations/langgraph-typescript/shared-tools/sales-todos.ts
/**
 * Sales todos tool implementation.
 *
 * TypeScript equivalent of showcase/shared/python/tools/sales_todos.py.
 */

import { SalesTodo } from "./types";

export const INITIAL_SALES_TODOS: SalesTodo[] = [
  {
    id: "st-001",
    title: "Follow up with Acme Corp on enterprise proposal",
    stage: "proposal",
    value: 85000,
    dueDate: "2026-04-15",
    assignee: "Sarah Chen",
    completed: false,
  },
  {
    id: "st-002",
    title: "Qualify lead from TechFlow demo request",
    stage: "prospect",
    value: 42000,
    dueDate: "2026-04-18",
    assignee: "Mike Johnson",
    completed: false,
  },
  {
    id: "st-003",
    title: "Send contract to DataViz Inc for final review",
    stage: "negotiation",
    value: 120000,
    dueDate: "2026-04-20",
    assignee: "Sarah Chen",
    completed: false,
  },
];

/**
 * Assign crypto.randomUUID() to any todos missing an ID, then return
 * the updated list.
 */
export function manageSalesTodosImpl(todos: Partial<SalesTodo>[]): SalesTodo[] {
  return todos.map((todo) => ({
    id: todo.id || crypto.randomUUID(),
    title: todo.title ?? "",
    stage: todo.stage ?? "prospect",
    value: todo.value ?? 0,
    dueDate: todo.dueDate ?? "",
    assignee: todo.assignee ?? "",
    completed: todo.completed ?? false,
  }));
}

/**
 * Return current todos or initial defaults if none provided.
 */
export function getSalesTodosImpl(
  currentTodos?: Partial<SalesTodo>[] | null,
): SalesTodo[] {
  if (currentTodos != null) {
    return currentTodos.length > 0 ? manageSalesTodosImpl(currentTodos) : [];
  }
  return [...INITIAL_SALES_TODOS];
}
